File: ioi/2016/day2/messy/solutions-to-submit/messy_java_subtask12.java.dir/messy.java
// name = messy_java_subtask12.java, type = java7

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.Writer;
import java.util.Arrays;
import java.util.HashSet;
import java.util.InputMismatchException;
import java.util.Set;

/**
 * Built using CHelper plug-in
 * Actual solution is at the top
 *
 * @author Egor Kulikov (egor@egork.net)
 */



class messy {
    public int[] restore_permutation(int n, int w, int r) {
        char[] element = new char[n];
        Arrays.fill(element, '0');
        for (int i = 0; i < n; i++) {
            element[i] = '1';
            grader.add_element(new String(element));
        }
        grader.compile_set();
        int[] answer = new int[n];
        boolean[] was = new boolean[n];
        Arrays.fill(element, '0');
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (was[j]) {
                    continue;
                }
                element[j] = '1';
                if (grader.check_element(new String(element))) {
                    answer[j] = i;
                    was[j] = true;
                    break;
                }
                element[j] = '0';
            }
        }
        return answer;
    }

}

public class messy_java_subtask12 {
    public static void main(String[] args) {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        OutputWriter out = new OutputWriter(outputStream);
        subtask123 solver = new subtask123();
        solver.solve(1, in, out);
        out.close();
    }

    static class subtask123 {
        public void solve(int testNumber, InputReader in, OutputWriter out) {
            
        }

    }

    static class InputReader {
        private InputStream stream;
        private byte[] buf = new byte[1024];
        private int curChar;
        private int numChars;
        private InputReader.SpaceCharFilter filter;

        public InputReader(InputStream stream) {
            this.stream = stream;
        }

        public int read() {
            if (numChars == -1) {
                throw new InputMismatchException();
            }
            if (curChar >= numChars) {
                curChar = 0;
                try {
                    numChars = stream.read(buf);
                } catch (IOException e) {
                    throw new InputMismatchException();
                }
                if (numChars <= 0) {
                    return -1;
                }
            }
            return buf[curChar++];
        }

        public int readInt() {
            int c = read();
            while (isSpaceChar(c)) {
                c = read();
            }
            int sgn = 1;
            if (c == '-') {
                sgn = -1;
                c = read();
            }
            int res = 0;
            do {
                if (c < '0' || c > '9') {
                    throw new InputMismatchException();
                }
                res *= 10;
                res += c - '0';
                c = read();
            } while (!isSpaceChar(c));
            return res * sgn;
        }

        public boolean isSpaceChar(int c) {
            if (filter != null) {
                return filter.isSpaceChar(c);
            }
            return isWhitespace(c);
        }

        public static boolean isWhitespace(int c) {
            return c == ' ' || c == '\n' || c == '\r' || c == '\t' || c == -1;
        }

        public interface SpaceCharFilter {
            public boolean isSpaceChar(int ch);

        }

    }

    static class OutputWriter {
        private final PrintWriter writer;

        public OutputWriter(OutputStream outputStream) {
            writer = new PrintWriter(new BufferedWriter(new OutputStreamWriter(outputStream)));
        }

        public OutputWriter(Writer writer) {
            this.writer = new PrintWriter(writer);
        }

        public void print(Object... objects) {
            for (int i = 0; i < objects.length; i++) {
                if (i != 0) {
                    writer.print(' ');
                }
                writer.print(objects[i]);
            }
        }

        public void print(int[] array) {
            for (int i = 0; i < array.length; i++) {
                if (i != 0) {
                    writer.print(' ');
                }
                writer.print(array[i]);
            }
        }

        public void printLine(int[] array) {
            print(array);
            writer.println();
        }

        public void printLine(Object... objects) {
            print(objects);
            writer.println();
        }

        public void close() {
            writer.close();
        }

    }
}

File: ioi/2016/day2/messy/solutions-to-submit/messy_iz.cpp.dir/messy.cpp
// name = messy_iz.cpp, type = cpp.g++11

#include "messy.h"


#include <vector>
#include <string>

//#include "messy.h"

using namespace std;

std::vector<int> restore_permutation(int n, int w, int r) {
    for (int len = n / 2; len >= 1; len >>= 1) {
        for (int s = 0; s < n; s += 2 * len) {
            string a = "";
            for (int i = 0; i < n; i++) a += '1';
            for (int i = 0; i < len * 2; i++) a[s + i] = '0';
            for (int i = 0; i < len; i++) {
                string ss = a;
                ss[s + i] = '1';
                add_element(ss);
            }
        }
    }
    compile_set();
    vector<vector<int> > can(n);
    for (int i = 0; i < n; i++) can[0].push_back(i);
    for (int len = n / 2; len >= 1; len >>= 1) {
        for (int s = 0; s < n; s += 2 * len) {
            string a = "";
            for (int i = 0; i < n; i++) a += '1';
            for (int i = 0; i < (int)can[s].size(); i++) a[can[s][i]] = '0';
            vector<int> c1, c2;
            for (int i = 0; i < (int)can[s].size(); i++) {
                string ss = a;
                ss[can[s][i]] = '1';
                if (check_element(ss)) {
                    c1.push_back(can[s][i]);
                } else {
                    c2.push_back(can[s][i]);
                }
            }
            can[s] = c1;
            can[s + len] = c2;
        }
    }
    vector<int> ans(n);
    for (int i = 0; i < n; i++) ans[can[i][0]] = i;
    return ans;
}
